File: Yr2022/day_five.py
from time import perf_counter_ns

from queue import LifoQueue



########## LOGIC USING STACKS

def put_stack_from_line(stack_list, input_line):
    for char_idx, char in enumerate(input_line):
        if char not in [' ', '[',']', '\n']:
            stack_list[(char_idx - 1)//4].put(char)

def initialise_stacks(stack_initialiser_q):
    stack_list = []
    for character in stack_initialiser_q.get():
        if character.isdigit():
            stack_list.append(LifoQueue())
    while not stack_initialiser_q.empty():
        put_stack_from_line(stack_list, stack_initialiser_q.get())
    return stack_list


def shift_single_crate(stack_list, origin_idx, dest_idx):
    stack_list[dest_idx].put(stack_list[origin_idx].get())


def shift_multi_crates(stack_list, origin_idx, dest_idx, num_crates):
    for i in range(0,num_crates):
        shift_single_crate(stack_list, origin_idx, dest_idx)


def shift_multi_crates_no_reverse(stack_list, origin_idx, dest_idx, num_crates):
    transfer_list = []
    for i in range(0,num_crates):
        transfer_list.append(stack_list[origin_idx].get())
    for item in reversed(transfer_list):
        stack_list[dest_idx].put(item)


def calculate_tops_stacks(filename, reverse=False):
    stack_initialiser_q = LifoQueue()
    initialising_stacks = True
    stack_list = []    
    with open(filename, 'r') as file:        
        for line in file:
            if initialising_stacks:
                if line.strip() == "":
                    initialising_stacks = False
                    stack_list = initialise_stacks(stack_initialiser_q)
                else:
                    stack_initialiser_q.put(line)
            
            else:
                line_break = [int(x) for x in line.strip().split(' ') if x.isdigit()]
                if reverse:
                    shift_multi_crates(stack_list, line_break[1]-1, line_break[2]-1, line_break[0])
                else:
                    shift_multi_crates_no_reverse(stack_list, line_break[1]-1, line_break[2]-1, line_break[0])

    result_list = []
    for stack in stack_list:
        result_list.append(stack.get())            
    return result_list



########## LOGIC USING STRINGS



def initialise_strings(string_initialiser_list):
    max_stack_idx = int(string_initialiser_list[-1].strip()[-1])
    result_string_list = ["" for x in range(0, max_stack_idx)]
    for line in reversed(string_initialiser_list[0:-1]):
        for i in range(0, max_stack_idx):
            char_idx = i*4 + 1
            if line[char_idx] != ' ':
                result_string_list[i] += line[char_idx]
    return result_string_list
            

def transfer_strings(string_list, origin_idx, dest_idx, num_crates, reverse=False):
    string_list[origin_idx], to_transfer = string_list[origin_idx][0:-num_crates], string_list[origin_idx][-num_crates:]
    transfer_list = to_transfer[::-1] if reverse else to_transfer
    string_list[dest_idx] += transfer_list


def calculate_tops_strings(filename, reverse=False):
    initialising_stacks = True
    string_initialiser_list = []
    string_list = []
    with open(filename, 'r') as file:        
        for line in file:
            if initialising_stacks:
                if line.strip() == "":
                    initialising_stacks = False
                    string_list = initialise_strings(string_initialiser_list)
                else:
                    string_initialiser_list.append(line)
            else:
                line_break = [int(x) for x in line.strip().split(' ') if x.isdigit()]              
                transfer_strings(string_list, line_break[1]-1, line_break[2]-1, line_break[0], reverse)
    return [each_string[-1] for each_string in string_list]



if __name__ == "__main__":
    time_start = perf_counter_ns()
    file_loc = r"C:\Users\Chris\Documents\CodeProjects\AdventOfCode\Yr2022\PuzzleData\puzzle_day_five.txt"
    reverse = False
    #print(calculate_tops_stacks(file_loc, reverse))    
    print(calculate_tops_strings(file_loc, reverse))    

    print(f"Elapsed time : {(perf_counter_ns() - time_start)/1000:.0f} microseconds")

File: Yr2022/src/day_five.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use std::time::Instant;




fn main() {   
    let start = Instant::now();
    let file_name= "C:\\Users\\Chris\\Documents\\CodeProjects\\AdventOfCode\\Yr2022\\PuzzleData\\puzzle_day_five.txt";
    let reverse: bool = false;
    let tops = calculate_tops_stacks(file_name, reverse);
    //let tops = calculate_tops_strings(file_loc, reverse));
    
    let duration = start.elapsed();
    println!("Top items : {:?}", tops);    
    println!("Time taken to execute: {} microseconds", duration.as_micros());    
}


fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}


fn initialise_stacks(stack_initialiser_q: &Vec<String> ) -> Vec<Vec<String>> {
    let mut stack_vec : Vec<Vec<String>> = Vec::new();
    let max_stack_idx_str: &str = stack_initialiser_q.last().unwrap().trim().split(" ").last().unwrap();
    let max_stack_idx : usize = max_stack_idx_str.parse::<usize>().unwrap();    

    for _vec_idx in 0..max_stack_idx as usize {
        stack_vec.push(Vec::new());
    }

    for row in stack_initialiser_q.iter().rev().skip(1) {
        let this_row: Vec<&str> = row.split("] ")        
                                        .collect();
                   
        let mut skipped_entries : usize = 0;
        for (idx, item) in this_row.iter().enumerate(){            
            if item.trim().is_empty() {continue;}
            let pre_suff : Vec<&str> = item.split("[").collect();
            skipped_entries += pre_suff[0].len()/4;
            let stack_idx = idx + skipped_entries as usize;
            stack_vec[stack_idx].push(pre_suff[1].to_string().replace("]",""));            
        }
    }
    return stack_vec;
}



fn shift_crates(stack_vec: &mut Vec<Vec<String>>, origin_idx: usize, dest_idx:usize,
                 num_crates: usize, reverse:bool) {
    let mut transfer_vec : Vec<String> = Vec::new();
    for _i in 0..num_crates {
        transfer_vec.push(stack_vec[origin_idx].pop().unwrap());        
    }

    if reverse {
        for element in transfer_vec.into_iter().rev() {
            stack_vec[dest_idx].push(element);
        }
    }
    else {
        for element in transfer_vec {
            stack_vec[dest_idx].push(element);
        }
    }
    
}


fn get_tops(stack_vec: Vec<Vec<String>>) -> String {
    let mut result: String = String::new();
    for stack in stack_vec {
        result.push_str(stack.last().unwrap());
    }
    return result;
}



fn calculate_tops_stacks(my_file_name: &str, reverse:bool) -> String {
    let mut initialising_stacks: bool = true;
    let mut stack_initialiser_q: Vec<String> = Vec::new();
    let mut stack_vec: Vec<Vec<String>> = Vec::new();
    
    if let Ok(lines) = read_lines(my_file_name) {
        for line in lines {
            if let Ok(ip) = line {                
                if initialising_stacks {
                    //println!("{}", ip);
                    if ip.is_empty() {
                        initialising_stacks = false;
                        stack_vec = initialise_stacks(&stack_initialiser_q);
                    }
                    else {
                        stack_initialiser_q.push(ip.clone());
                    }
                }

                else {

                    let command_vec : Vec<&str> = ip.split(' ').collect();
                    if command_vec[0] == "move" {
                        shift_crates(&mut stack_vec, command_vec[3].parse::<usize>().unwrap()-1,
                                    command_vec[5].parse::<usize>().unwrap()-1,
                                    command_vec[1].parse::<usize>().unwrap(),
                                    reverse);
                    }
                    //println!("COMMAND {:?}", command_vec);

                }             
            }
        }
    }

    //let tops : Vec<&str> = Vec::new();
    //return tops;

    /*
    println!("-------------------------------------");
    for element in &stack_vec {
        println!("{:?}", element.last());

    }
    println!("-------------------------------------");
    */ 

    //println!("{:?}", stack_vec);

    //return stack_initialiser_q;
    return get_tops(stack_vec);
}
